### PR TITLE
Add method to add instructions to a DAGCircuit from an iterator of PackedInstruction

### LINKED ISSUE #13003: Create a `DAGCircuit::add_from_iter()` method in rust
### What should we add?

Tracked by #13001

Create a method `DAGCircuit::add_from_iter` to create add a chain of `PackedInstruction` instances to a `DAGCircuit` preventing the duplicate insertion of `DAGOutputNode` instances.

### PR DESCRIPTION
**Pull request by raynelfss**
Solves #13003 

### Summary
Tracked by #13001 and preceeded by #12975
The following commits aim to add a `DAGCircuit::extend()` to add a chain of `DAGOpNodes` to the DAGCircuit based on a sequence of `PackedInstruction`.

### Details and comments
These commits add the following two methods:
- `DAGCircuit::extend()` to add a sequence of `PackedInstruction` continously avoiding having to re-add each bit/var's output node each time an instruction is added. The addition of the output node is only performed once all instructions have been added correctly to the DAGCircuit.

### Known issues:
- At the moment, I'm still thinking of ways of handling vars correctly as we need to keep track of each bit/var's output node and vars are currently represented as `PyObject` which are not hashable.
    - ~~One possible solution would be to use a `PyDict` however the structures that are being stored within this mapping are rust-native (`NodeIndex`, `Wire`)~~.
    - ~~Another solution would be to somehow internalize these structures, which I'm still thinking on how to do.~~
- Feel free to comment with any ideas.

### Blockers
- [x] #12975 (will rebase once merged)
- [x] #12550

### CODE REVIEW

**Comment by ElePT** [comment]
I believe that the PR doesn't have any open blocker. I will remove the on hold label. Let me know if I missed something.

**Comment by raynelfss** [comment]
> I believe that the PR doesn't have any open blocker. I will remove the on hold label. Let me know if I missed something.

I was about to do that :) Thank you

File: crates/circuit/src/dag_circuit.rs
Comment (ElePT) [inline crates/circuit/src/dag_circuit.rs]: If I am not mistaken it looks like you were able to store the var keys as PyDicts, right? Is there something else to fix or was the TODO addressed?

File: crates/circuit/src/dag_circuit.rs
Comment (ElePT) [inline crates/circuit/src/dag_circuit.rs]:
```suggestion
    /// Extends the DAG with valid instances of [PackedInstruction]
    pub fn extend<I>(&mut self, py: Python, iter: I) -> PyResult<Vec<NodeIndex>>
```

**Review by ElePT — COMMENTED** [review]
Thanks @raynelfss! This PR is quite useful for transpiler passes that build new instructions (like unitary synthesis). I have been using these changes in my PR and everything has worked as expected so far, so you can have that as a "unit test" data point. I might be overseeing something but I would approve this PR after addressing a few minor (and quite superficial) comments.

File: crates/circuit/src/dag_circuit.rs
Comment (raynelfss) [reply crates/circuit/src/dag_circuit.rs]: You are correct, I must have forgotten to remove it.

**Review by ElePT — DISMISSED** [review]
LGTM

**Review by ElePT — COMMENTED** [review]
Actually, I think I have found an issue when trying to add instructions that have no clbits. When I initialize a packed instruction with `clbits: target_dag.cargs_interner.get_default()`,  I then get a `the caller is responsible for only using interner keys from the correct interner` error. I will try to come up with a minimal example that reproduces the problem.

**Comment by ElePT** [comment]
The issue was fixed by #13092, it was unrelated to the changes in this PR :)

**Review by ElePT — DISMISSED** [review]

**Comment by kevinhartman** [comment]
I can give this a quick look as well if it's helpful (I see it is not in the merge queue yet)

File: crates/circuit/src/dag_circuit.rs
Comment (kevinhartman) [inline crates/circuit/src/dag_circuit.rs]: As an optimization, you can use `iter.size_hint()` here to reserve an appropriately sized vector if whatever was passed for `iter` happens to provide `Some(upper_bound)` in the tuple returned by `size_hint`.

File: crates/circuit/src/dag_circuit.rs
Comment (kevinhartman) [inline crates/circuit/src/dag_circuit.rs:6398]: Can you add a comment here to explain that what this code is doing is removing the output node edge the first time a qubit is used in `iter`, and that we re-add them after processing all instructions?

It also may be cleaner to use the `entry` API up on line 6393 to get a mutable reference to the value instead of removing `qubit` from `qubit_last_nodes` only to add it back with a different value, but I think a comment would at least be helpful.

**Review by kevinhartman — COMMENTED** [review]
This is looking good, thanks for adding this @raynelfss!

The one larger point I think we should start to be weary of is introducing new methods that aren't covered by any of our existing testing. We've operated in a bit of a free-for-all the last few months while porting things to Rust, under the assumption that our existing Python test code would provide sufficient coverage. 

IMO, this new method is sufficiently complex and should be unit tested. I'm not sure if we've decided on an approach to exercise the Rust-only API of our `pyclass`s. Perhaps @mtreinish has opinions / suggestions?

**Comment by mtreinish** [comment]
> IMO, this new method is sufficiently complex and should be unit tested. I'm not sure if we've decided on an approach to exercise the Rust-only API of our pyclasss. Perhaps @mtreinish has opinions / suggestions?

In general having rust unit tests is possible and we have them already in place where it is possible/makes sense for PRs like this with new rust specific functionality. See: https://github.com/Qiskit/qiskit/commit/86c63eb3d719a129dbcb92c56e9a67954030b803 for a recent example (which is already setup to run in CI: https://dev.azure.com/qiskit-ci/qiskit/_build/results?buildId=60452&view=logs&j=b9878877-bc64-52fc-fc03-577dd0aa7cb4&t=2cc4852e-f347-585e-188c-b93a0bf20b0f&l=363 ). But what we will struggle with here is that this new method takes a `py` token and we don't have a mechanism to interact with python via rust unit tests.

I would agree we should have this exercised somehow in an ideal world. But I think we probably are ok to let this slide if we can't come up with a solution for unit testing with a python dependency. Given that there are several PRs actively based on top of this and those are exercising the code and tested via Python AFAICT.

File: crates/circuit/src/dag_circuit.rs
Comment (raynelfss) [reply crates/circuit/src/dag_circuit.rs]: In this case, since we don't exactly know the type of iterator the received collection will turn into, we can't obtain a size hint until we consume the collection into the iterator. This should be fine as long as we use the iterator right after.

**Review by kevinhartman — APPROVED** [review]
LGTM, thanks for the changes!

**Pull request merged**
